## Description: Adding two signals that are needed for Application Indicator to
##              keep up-to-date with changed GtkMenu(Item)s.
## Origin/Author: Cody Russell <bratsche@gnome.org>, 
##                Sense Hofstede <sense@ubuntu.com>
## Bug: unknown and https://launchpad.net/bugs/608219
Index: gtk+-2.91.7/gtk/gtkcontainer.c
===================================================================
--- gtk+-2.91.7.orig/gtk/gtkcontainer.c	2010-12-22 03:36:24.000000000 +1100
+++ gtk+-2.91.7/gtk/gtkcontainer.c	2010-12-22 13:23:21.983372001 +1100
@@ -244,6 +244,8 @@
   REMOVE,
   CHECK_RESIZE,
   SET_FOCUS_CHILD,
+  CHILD_ADDED,
+  CHILD_REMOVED,
   LAST_SIGNAL
 };
 
@@ -504,6 +506,24 @@
 		  _gtk_marshal_VOID__OBJECT,
 		  G_TYPE_NONE, 1,
 		  GTK_TYPE_WIDGET);
+  container_signals[CHILD_ADDED] =
+    g_signal_new (I_("child-added"),
+                  G_OBJECT_CLASS_TYPE (gobject_class),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL, NULL,
+                  _gtk_marshal_VOID__OBJECT,
+                  G_TYPE_NONE, 1,
+                  GTK_TYPE_WIDGET);
+  container_signals[CHILD_REMOVED] =
+    g_signal_new (I_("child-removed"),
+                  G_OBJECT_CLASS_TYPE (gobject_class),
+                  G_SIGNAL_RUN_LAST,
+                  0,
+                  NULL, NULL,
+                  _gtk_marshal_VOID__OBJECT,
+                  G_TYPE_NONE, 1,
+                  GTK_TYPE_WIDGET);
 
   g_type_class_add_private (class, sizeof (GtkContainerPrivate));
 }
Index: gtk+-2.91.7/gtk/gtkmenuitem.c
===================================================================
--- gtk+-2.91.7.orig/gtk/gtkmenuitem.c	2010-12-22 13:23:11.843372001 +1100
+++ gtk+-2.91.7/gtk/gtkmenuitem.c	2010-12-22 13:23:21.983372001 +1100
@@ -53,6 +53,8 @@
   TOGGLE_SIZE_ALLOCATE,
   SELECT,
   DESELECT,
+  SUBMENU_ADDED,
+  SUBMENU_REMOVED,
   LAST_SIGNAL
 };
 
@@ -271,6 +273,25 @@
                   _gtk_marshal_VOID__VOID,
                   G_TYPE_NONE, 0);
 
+  menu_item_signals[SUBMENU_ADDED] =
+    g_signal_new (I_("submenu-added"),
+		  G_OBJECT_CLASS_TYPE (gobject_class),
+		  G_SIGNAL_RUN_FIRST,
+ 		  0,
+		  NULL, NULL,
+		  _gtk_marshal_VOID__OBJECT,
+		  G_TYPE_NONE, 1,
+		  GTK_TYPE_WIDGET);
+ 
+ menu_item_signals[SUBMENU_REMOVED] =
+    g_signal_new (I_("submenu-removed"),
+		  G_OBJECT_CLASS_TYPE (gobject_class),
+		  G_SIGNAL_RUN_FIRST,
+ 		  0,
+		  NULL, NULL,
+		  _gtk_marshal_VOID__VOID,
+		  G_TYPE_NONE, 0);
+
   /**
    * GtkMenuItem:right-justified:
    *
@@ -1242,6 +1265,11 @@
       if (gtk_widget_get_parent (GTK_WIDGET (menu_item)))
         gtk_widget_queue_resize (GTK_WIDGET (menu_item));
 
+      if(submenu == NULL)
+        g_signal_emit_by_name (menu_item, "submenu-removed");
+      else
+        g_signal_emit_by_name (menu_item, "submenu-added", GTK_WIDGET(submenu));
+
       g_object_notify (G_OBJECT (menu_item), "submenu");
     }
 }
Index: gtk+-2.91.7/gtk/gtkmenushell.c
===================================================================
--- gtk+-2.91.7.orig/gtk/gtkmenushell.c	2010-12-22 13:23:11.843372001 +1100
+++ gtk+-2.91.7/gtk/gtkmenushell.c	2010-12-22 13:23:21.983372001 +1100
@@ -571,6 +571,8 @@
                               GTK_WIDGET (menu_shell),
                               child,
                               position);
+
+  g_signal_emit_by_name (menu_shell, "child-added", child);
 }
 
 static void
